### DISCUSSION TITLE
Alternatives to data tiddlers?

### DISCUSSION THREAD

**Discussion by twMat** [TiddlyWiki/TiddlyWiki5#5634] [General]
@Jermolene has explained that he wishes to replace data tiddlers. Yet they are popular, perhaps mostly because of their ease of use. 

I want to see if we can come up with "design ideas" for a good replacement for data tiddlers. Deprecating data tiddlers is thus a presumed starting point in this discussion, so the discussion is **not** about whether data tiddlers should stay or not, or if they're good or not.

...

One (obvious?) idea might be to use fields to define the key-value pairs. 

For this, we'd need a different edit template for this to be practical because it would be slow to feed in one field at a time and click [create field] etc. If the `text` field is still there, it would be possible to keep notes about the dictionary, which would be nice.

If individual fields define the key-value pairs, then we'd probably need some mechanism to exclude (black list) specific fields from being treated as such pairs. (I can't recall if there is some kind of "get all indexes" tool (filter op?) but this is a case where this woudl be a concern). Or maybe a white list.

One idea would be to enter a special "edit mode" when editing data tiddlers that converts the fields into one easily editable text field (resembling how they currently look) and when saving the tiddler they are converted back into fields. This might be a backward compatible solution? In addition to the automatic conversion, there could be a button to manually toggle back and forth. This is not super elegant UI/UX but data tiddlers is probably not a daily thing to edit manually so it'd be sufficient IMO.

Another UI idea is to make edit mode custom fields more accessible. Tiddler type "data tiddler" could present the fields as a two column table (think Excel) and one could easily navigate with arrow buttons or maybe tab button. I like this idea. Should make for a very clean UI.

...apropos Excel... could it even allow calculated values depending on other index values? Does this thought even make sense?

|foo|3|
|bar|5|
|sum|foo+bar|

...

A presumption here is that fields indeed would be an acceptible solution. Another idea would be a separate tiddler for each key-value pair. Sounds bloaty to me, but who knows.

Thoughts? Further ideas?

**Comment by pmario** [comment]
Just a thought: For political correctness we could use "blocklist" and "passlist" instead of black&white

**Reply by twMat** [reply]
Is that established terminology for what it refers to?

**Comment by pmario** [comment]
I think we should define what are the properties that make datatiddlers useful and "easy to use". Then we can think about a "replacement that has the same properties" and more. 

 - using datatiddlers "feels" easy / good
   - If you export / import a datatiddler you can be sure "you get it all"
 - can be edited with the default text editor
 - Copy / Paste is possible
 - 1 tiddler can transport many settings
 - easy to store log entries / time-line
 - View and edit workflow is fast
 - can be made "self sufficient" 
   - no meta data needed eg: no blocklists
 - is a very basic "key : value" store
   - "Key - Value" store is a "buzz word"
 - ...

That are my minimal properties that need to be satisfied. Or they should better be excelled!

**Reply by twMat** [reply]
Not a must have, but a nice have:

Distinct columns for easy reading. The current syntax makes the value column "uneven".

**Comment by dubiouscript** [comment]
i remember previously reading about this 

> @Jermolene has explained that he wishes to replace data tiddlers.

... managed to find this comment 

https://github.com/Jermolene/TiddlyWiki5/pull/4389#issuecomment-613557072

> I don't think we're quite at the point that we want to broadly discourage users from using data tiddlers because there are still quite a few use cases where they are the best option. The situation is more that I want to discourage further development of data tiddler features, and instead try to focus on better ways of achieving those remaining use cases.

wrt alternatives i found this mention of 'JSON Pointers to reference data'

https://github.com/Jermolene/TiddlyWiki5/issues/411

but tbh idk how that compares to whats curetly posible with data tidlers methods ect ..

**Reply by twMat** [reply]
>The situation is more that I want to discourage further development of data tiddler features, and instead try to focus on better ways of achieving those remaining use cases.

The problem is: Unless those "better ways" are identified, this whole thing is stuck in limbo.

Perhaps the new JSON storage changes the premises for the concept of data tiddlers?

**Reply by pmario** [reply]
>Perhaps the new JSON storage changes the premises for the concept of data tiddlers?

edited:
~~I think it doesn't~~. It doesn't change how data-tiddler refresh is handled, as Joshua describes it.

**Reply by pmario** [reply]
Second thought - It may be, but we need a better EditTemplate

See: **How the TW internal data structure looks like and why data-tiddlers are not optimal** #6116

**Reply by twMat** [reply]
> See: **How the TW internal data structure looks like and why data-tiddlers are not optimal** #6116

That is one great writeup! Thank you very much for it. So *that* post clearly infers that we should use *fields*. So the question is how this would match your demands listed here above, to make datatiddlers "easy to use". 

Looking at your demands and thinking how this would work with fields, I'm thinking a **conversion** needs to take place when opening such a tiddler to edit it and when saving it. Basically, the edit "UI" can look like it does now, **all in the text field**, but when closing it, this is converted into separate fields. This conversion is likely slow but it is then fast to *use* the datatiddler.

AND there can be a neat viewmode UI showing a table with the field values and editors for the key values, similar to how custom fields are displayed, but here in viewmode. Then you can manually edit values without going into edit mode, i.e without the slow conversion.
(This viewmode UI would even resemble a *form*, something that we kind of miss in TW. )

For your demands:

> * If you export / import a datatiddler you can be sure "you get it all"

check

> * can be edited with the default text editor

check

> * Copy / Paste is possible

check

> * 1 tiddler can transport many settings

check (or maybe I don't understand what you mean with "transport settings")

> * easy to store log entries / time-line

?

> * View and edit workflow is fast

To view or edit is just like normal, but the very step in-between might not be. **But** we could even have a viewmode UI with direct editors to the field values! This makes the datatiddler pretty table! (Possibly even: For the instances where the key value is a *tiddler title* this could even show as links!) 

> * can be made "self sufficient"

check

> * no meta data needed eg: no blocklists

check

> * is a very basic "key : value" store

check

**Reply by pmario** [reply]
>Looking at your demands and thinking how this would work with fields, I'm thinking a conversion needs to take place when opening such a tiddler to edit it and when saving it. Basically, the edit "UI" can look like it does now, all in the text field, but when closing it, this is converted into separate fields. This conversion is likely slow but it is then fast to use the datatiddler.

Reading and writing data from the store to the UI is no performance problem, because it's always only 1 tiddler that is handled. Problems come up, if we need to read a lot of tiddlers, with "poorly" written filters and nested lists. ... 

There is a PR in the works, that will help us to create a better UI to edit fields: https://github.com/Jermolene/TiddlyWiki5/pull/6377 With the help of this PR we will be able to not only improve "date handling" but to improve the overall UI for fields. 

eg: It should be possible to have the same UI for fields as we have for tags now. ... It should be possible to create dynamic multiline editors for fields ... with some new templates. 

Those templates should also be available in view-mode if done right. .. So the end-result may be a UI that doesn't make a big difference between view and edit mode. ... The new "cascade" filters should have paved the way to those functions. ..

**Reply by twMat** [reply]
Oooh!  #6377 and what you describe gives me goose bumps! 

OK, so I take it that if #6377 is implemented then my idea here about "converting" text content into fields isn't really relevant then - and the overall problem with data tiddlers will likely be solved.

**Comment by joshuafontany** [comment]
The big issue is that the data is stored in the text field & this triggers all references in all widgets to update any time _any_ of the json data is updated. This skips over a large part of TW's rendering optimization. Imagine a data tiddler holding 100 "people" objects in an Array. Now open 50 tillders tagged "$:/person", each of which transcludes one of the data objects. Now change person-100's  email address. All 50 tiddlers rendered in the story river must be refreshed- even though visually the person changed might not even be displayed.


Now it would be REALLY COOL if, on "application/json" & other data tiddlers, that Field References would auto-grab the data (in the text field) at the same index...

I imagine calling tiddler.getFieldStrings() on a data tiddker & getting all the data as the "text" key but Also getting keys & values for all data objects in the text field... hrm then we have keyname clashes possible. Or maybe some other way to hook nested json data into the appropriate refresh mechanism.

**Reply by pmario** [reply]
@joshuafontany If I do open 50 data-tiddlers, I don't have a problem with some refresh-penalty. I expect it. ... IMO the big problem is if it touches the core UI. ...

**Reply by kookma** [reply]
> Now change person-100's email address. All 50 tiddlers rendered in the story river must be refreshed- even though visually the person changed might not even be displayed.

don't you think this is a design issue and should be addressed in the core?

**Reply by pmario** [reply]
> don't you think this is a design issue and should be addressed in the core?

This can only be addressed, if every element in a data-tiddler becomes a tiddler. ... But that's not what we (the users) want. 

Jeremy refused to merge the functionality that is now the [keyvalues](https://wikilabs.github.io/editions/keyvalues/)  plugin, because he doesn't want to add new functions for data-tiddlers to the core.

His opinion is, that it was a mistake to implement it in the first place and from a "code complexity" point of view he is right. 

There is a high chance, that the whole functionality will end up in a plugin and will be removed from the core. 

IMO the biggest "pro" for data-tiddlers is that users love them. ... So we will need a really good replacement to keep  everyone happy.

**Reply by saqimtiaz** [reply]
> The big issue is that the data is stored in the text field & this triggers all references in all widgets to update any time _any_ of the json data is updated. This skips over a large part of TW's rendering optimization. Imagine a data tiddler holding 100 "people" objects in an Array. Now open 50 tillders tagged "$:/person", each of which transcludes one of the data objects. Now change person-100's email address. All 50 tiddlers rendered in the story river must be refreshed- even though visually the person changed might not even be displayed.

Is this still the case in 5.2.0? I know transclusion now only refresh if the field being transcluded has changed rather than the tiddler. I don't recall if the same improvement works for indexes as well.

**Comment by pmario** [comment]
If you use data-tiddlers as a pure "key-value store" without a connection to the UI, it probably is the best way to deal with "single level" key-values. ... There is _no_ real alternative from my point of view.

**Reply by pmario** [reply]
There is some discussion going on about [typed fields](https://github.com/Jermolene/TiddlyWiki5/discussions/5805), with which it would be possible to create JSON-like tiddler fields. ... BUT ... In the end such a field would be a [mini-tiddler](https://github.com/Jermolene/TiddlyWiki5/discussions/5805#discussioncomment-920215) as Jeremy described it in a response in that thread. ... So it's a bit of a "dead end" ... kind of

**Comment by twMat** [comment]
Subthreads within the bigger thread have a tendency to hide so, as the momentarily bottom-most post, I just want to refer to @pmarios post [above](https://github.com/Jermolene/TiddlyWiki5/discussions/5634#discussioncomment-1977770) where he comments on #6377  . My interpretation is that this will solve the raised issue... which I now note is a Discussion, not an Issue, so it is not closeable. Thank you all.
